File: lib/services/tmdb_service.dart
import 'package:dio/dio.dart';
import '../utils/app_logger.dart';
import 'settings_service.dart';

/// TMDB API service for fetching trailers and additional metadata
class TmdbService {
  static const String _baseUrl = 'https://api.themoviedb.org/3';
  static TmdbService? _instance;

  String? _apiKey;
  final Dio _dio = Dio();

  TmdbService._();

  static Future<TmdbService> getInstance() async {
    if (_instance == null) {
      _instance = TmdbService._();
      await _instance!._init();
    }
    return _instance!;
  }

  Future<void> _init() async {
    final settings = await SettingsService.getInstance();
    _apiKey = settings.getTmdbApiKey();
  }

  /// Refresh API key from settings
  Future<void> refreshApiKey() async {
    final settings = await SettingsService.getInstance();
    _apiKey = settings.getTmdbApiKey();
  }

  bool get hasApiKey => _apiKey != null && _apiKey!.isNotEmpty;

  /// Search for a movie by title and year
  Future<int?> searchMovie(String title, {int? year}) async {
    if (!hasApiKey) return null;

    try {
      final response = await _dio.get(
        '$_baseUrl/search/movie',
        queryParameters: {
          'api_key': _apiKey,
          'query': title,
          if (year != null) 'year': year,
        },
      );

      if (response.statusCode == 200) {
        final results = response.data['results'] as List?;
        if (results != null && results.isNotEmpty) {
          return results.first['id'] as int?;
        }
      }
    } catch (e) {
      appLogger.w('TMDB movie search failed', error: e);
    }
    return null;
  }

  /// Search for a TV show by title
  Future<int?> searchTvShow(String title, {int? year}) async {
    if (!hasApiKey) return null;

    try {
      final response = await _dio.get(
        '$_baseUrl/search/tv',
        queryParameters: {
          'api_key': _apiKey,
          'query': title,
          if (year != null) 'first_air_date_year': year,
        },
      );

      if (response.statusCode == 200) {
        final results = response.data['results'] as List?;
        if (results != null && results.isNotEmpty) {
          return results.first['id'] as int?;
        }
      }
    } catch (e) {
      appLogger.w('TMDB TV search failed', error: e);
    }
    return null;
  }

  /// Get trailers for a movie
  Future<List<TrailerInfo>> getMovieTrailers(int movieId) async {
    if (!hasApiKey) return [];

    try {
      final response = await _dio.get(
        '$_baseUrl/movie/$movieId/videos',
        queryParameters: {'api_key': _apiKey},
      );

      if (response.statusCode == 200) {
        final results = response.data['results'] as List?;
        if (results != null) {
          return results
              .where((v) =>
                  v['site'] == 'YouTube' &&
                  (v['type'] == 'Trailer' || v['type'] == 'Teaser'))
              .map((v) => TrailerInfo.fromJson(v as Map<String, dynamic>))
              .toList();
        }
      }
    } catch (e) {
      appLogger.w('TMDB movie trailers failed', error: e);
    }
    return [];
  }

  /// Get trailers for a TV show
  Future<List<TrailerInfo>> getTvTrailers(int tvId) async {
    if (!hasApiKey) return [];

    try {
      final response = await _dio.get(
        '$_baseUrl/tv/$tvId/videos',
        queryParameters: {'api_key': _apiKey},
      );

      if (response.statusCode == 200) {
        final results = response.data['results'] as List?;
        if (results != null) {
          return results
              .where((v) =>
                  v['site'] == 'YouTube' &&
                  (v['type'] == 'Trailer' || v['type'] == 'Teaser'))
              .map((v) => TrailerInfo.fromJson(v as Map<String, dynamic>))
              .toList();
        }
      }
    } catch (e) {
      appLogger.w('TMDB TV trailers failed', error: e);
    }
    return [];
  }

  /// Get trailer for content by title
  Future<TrailerInfo?> getTrailerForTitle(
    String title, {
    bool isMovie = true,
    int? year,
  }) async {
    if (!hasApiKey) return null;

    try {
      final id = isMovie
          ? await searchMovie(title, year: year)
          : await searchTvShow(title, year: year);

      if (id == null) return null;

      final trailers = isMovie
          ? await getMovieTrailers(id)
          : await getTvTrailers(id);

      if (trailers.isNotEmpty) {
        // Prefer official trailers
        final official = trailers.where((t) => t.official).toList();
        return official.isNotEmpty ? official.first : trailers.first;
      }
    } catch (e) {
      appLogger.w('Failed to get trailer for $title', error: e);
    }
    return null;
  }

  /// Get full movie details including cast, ratings, and genres
  Future<TmdbMovieDetails?> getMovieDetails(int movieId) async {
    if (!hasApiKey) return null;

    try {
      final response = await _dio.get(
        '$_baseUrl/movie/$movieId',
        queryParameters: {
          'api_key': _apiKey,
          'append_to_response': 'credits,release_dates,external_ids',
        },
      );

      if (response.statusCode == 200) {
        return TmdbMovieDetails.fromJson(response.data);
      }
    } catch (e) {
      appLogger.w('TMDB movie details failed', error: e);
    }
    return null;
  }

  /// Get full TV show details including cast, ratings, and genres
  Future<TmdbTvDetails?> getTvDetails(int tvId) async {
    if (!hasApiKey) return null;

    try {
      final response = await _dio.get(
        '$_baseUrl/tv/$tvId',
        queryParameters: {
          'api_key': _apiKey,
          'append_to_response': 'credits,content_ratings,external_ids',
        },
      );

      if (response.statusCode == 200) {
        return TmdbTvDetails.fromJson(response.data);
      }
    } catch (e) {
      appLogger.w('TMDB TV details failed', error: e);
    }
    return null;
  }

  /// Get enriched metadata for content by title
  Future<TmdbEnrichedMetadata?> getEnrichedMetadata(
    String title, {
    bool isMovie = true,
    int? year,
  }) async {
    if (!hasApiKey) return null;

    try {
      final id = isMovie
          ? await searchMovie(title, year: year)
          : await searchTvShow(title, year: year);

      if (id == null) return null;

      if (isMovie) {
        final details = await getMovieDetails(id);
        final trailers = await getMovieTrailers(id);
        if (details != null) {
          return TmdbEnrichedMetadata(
            id: id,
            imdbId: details.imdbId,
            title: details.title,
            overview: details.overview,
            genres: details.genres,
            cast: details.cast.take(10).toList(),
            crew: details.crew.take(5).toList(),
            rating: details.contentRating,
            posterPath: details.posterPath,
            backdropPath: details.backdropPath,
            releaseDate: details.releaseDate,
            runtime: details.runtime,
            voteAverage: details.voteAverage,
            trailers: trailers,
          );
        }
      } else {
        final details = await getTvDetails(id);
        final trailers = await getTvTrailers(id);
        if (details != null) {
          return TmdbEnrichedMetadata(
            id: id,
            imdbId: details.imdbId,
            title: details.name,
            overview: details.overview,
            genres: details.genres,
            cast: details.cast.take(10).toList(),
            crew: details.crew.take(5).toList(),
            rating: details.contentRating,
            posterPath: details.posterPath,
            backdropPath: details.backdropPath,
            releaseDate: details.firstAirDate,
            runtime: details.episodeRunTime.isNotEmpty ? details.episodeRunTime.first : null,
            voteAverage: details.voteAverage,
            trailers: trailers,
          );
        }
      }
    } catch (e) {
      appLogger.w('Failed to get enriched metadata for $title', error: e);
    }
    return null;
  }
}

/// Information about a trailer
class TrailerInfo {
  final String id;
  final String key;
  final String name;
  final String site;
  final String type;
  final bool official;

  TrailerInfo({
    required this.id,
    required this.key,
    required this.name,
    required this.site,
    required this.type,
    required this.official,
  });

  factory TrailerInfo.fromJson(Map<String, dynamic> json) {
    return TrailerInfo(
      id: json['id']?.toString() ?? '',
      key: json['key'] ?? '',
      name: json['name'] ?? '',
      site: json['site'] ?? '',
      type: json['type'] ?? '',
      official: json['official'] ?? false,
    );
  }

  /// Get YouTube thumbnail URL
  String get thumbnailUrl => 'https://img.youtube.com/vi/$key/hqdefault.jpg';

  /// Get YouTube embed URL
  String get embedUrl => 'https://www.youtube.com/embed/$key?autoplay=1';

  /// Get YouTube watch URL
  String get watchUrl => 'https://www.youtube.com/watch?v=$key';
}

/// Movie details from TMDB
class TmdbMovieDetails {
  final int id;
  final String title;
  final String? overview;
  final String? posterPath;
  final String? backdropPath;
  final String? releaseDate;
  final int? runtime;
  final double? voteAverage;
  final String? imdbId;
  final List<String> genres;
  final List<TmdbCastMember> cast;
  final List<TmdbCrewMember> crew;
  final String? contentRating;

  TmdbMovieDetails({
    required this.id,
    required this.title,
    this.overview,
    this.posterPath,
    this.backdropPath,
    this.releaseDate,
    this.runtime,
    this.voteAverage,
    this.imdbId,
    this.genres = const [],
    this.cast = const [],
    this.crew = const [],
    this.contentRating,
  });

  factory TmdbMovieDetails.fromJson(Map<String, dynamic> json) {
    // Extract genres
    final genreList = (json['genres'] as List?)
        ?.map((g) => g['name'] as String)
        .toList() ?? [];

    // Extract cast
    final credits = json['credits'] as Map<String, dynamic>?;
    final castList = (credits?['cast'] as List?)
        ?.map((c) => TmdbCastMember.fromJson(c))
        .toList() ?? [];
    final crewList = (credits?['crew'] as List?)
        ?.map((c) => TmdbCrewMember.fromJson(c))
        .toList() ?? [];

    // Extract US content rating
    String? rating;
    final releaseDates = json['release_dates'] as Map<String, dynamic>?;
    if (releaseDates != null) {
      final results = releaseDates['results'] as List?;
      if (results != null) {
        for (final region in results) {
          if (region['iso_3166_1'] == 'US') {
            final releases = region['release_dates'] as List?;
            if (releases != null && releases.isNotEmpty) {
              rating = releases.first['certification'] as String?;
              break;
            }
          }
        }
      }
    }

    // Extract IMDB ID
    final externalIds = json['external_ids'] as Map<String, dynamic>?;
    final imdbId = externalIds?['imdb_id'] as String?;

    return TmdbMovieDetails(
      id: json['id'] ?? 0,
      title: json['title'] ?? '',
      overview: json['overview'],
      posterPath: json['poster_path'],
      backdropPath: json['backdrop_path'],
      releaseDate: json['release_date'],
      runtime: json['runtime'],
      voteAverage: (json['vote_average'] as num?)?.toDouble(),
      imdbId: imdbId,
      genres: genreList,
      cast: castList,
      crew: crewList,
      contentRating: rating,
    );
  }

  String? get posterUrl => posterPath != null
      ? 'https://image.tmdb.org/t/p/w500$posterPath'
      : null;

  String? get backdropUrl => backdropPath != null
      ? 'https://image.tmdb.org/t/p/original$backdropPath'
      : null;
}

/// TV show details from TMDB
class TmdbTvDetails {
  final int id;
  final String name;
  final String? overview;
  final String? posterPath;
  final String? backdropPath;
  final String? firstAirDate;
  final List<int> episodeRunTime;
  final double? voteAverage;
  final String? imdbId;
  final List<String> genres;
  final List<TmdbCastMember> cast;
  final List<TmdbCrewMember> crew;
  final String? contentRating;

  TmdbTvDetails({
    required this.id,
    required this.name,
    this.overview,
    this.posterPath,
    this.backdropPath,
    this.firstAirDate,
    this.episodeRunTime = const [],
    this.voteAverage,
    this.imdbId,
    this.genres = const [],
    this.cast = const [],
    this.crew = const [],
    this.contentRating,
  });

  factory TmdbTvDetails.fromJson(Map<String, dynamic> json) {
    // Extract genres
    final genreList = (json['genres'] as List?)
        ?.map((g) => g['name'] as String)
        .toList() ?? [];

    // Extract cast
    final credits = json['credits'] as Map<String, dynamic>?;
    final castList = (credits?['cast'] as List?)
        ?.map((c) => TmdbCastMember.fromJson(c))
        .toList() ?? [];
    final crewList = (credits?['crew'] as List?)
        ?.map((c) => TmdbCrewMember.fromJson(c))
        .toList() ?? [];

    // Extract US content rating
    String? rating;
    final contentRatings = json['content_ratings'] as Map<String, dynamic>?;
    if (contentRatings != null) {
      final results = contentRatings['results'] as List?;
      if (results != null) {
        for (final region in results) {
          if (region['iso_3166_1'] == 'US') {
            rating = region['rating'] as String?;
            break;
          }
        }
      }
    }

    // Extract IMDB ID
    final externalIds = json['external_ids'] as Map<String, dynamic>?;
    final imdbId = externalIds?['imdb_id'] as String?;

    return TmdbTvDetails(
      id: json['id'] ?? 0,
      name: json['name'] ?? '',
      overview: json['overview'],
      posterPath: json['poster_path'],
      backdropPath: json['backdrop_path'],
      firstAirDate: json['first_air_date'],
      episodeRunTime: (json['episode_run_time'] as List?)
          ?.map((e) => e as int)
          .toList() ?? [],
      voteAverage: (json['vote_average'] as num?)?.toDouble(),
      imdbId: imdbId,
      genres: genreList,
      cast: castList,
      crew: crewList,
      contentRating: rating,
    );
  }

  String? get posterUrl => posterPath != null
      ? 'https://image.tmdb.org/t/p/w500$posterPath'
      : null;

  String? get backdropUrl => backdropPath != null
      ? 'https://image.tmdb.org/t/p/original$backdropPath'
      : null;
}

/// Cast member from TMDB credits
class TmdbCastMember {
  final int id;
  final String name;
  final String? character;
  final String? profilePath;

  TmdbCastMember({
    required this.id,
    required this.name,
    this.character,
    this.profilePath,
  });

  factory TmdbCastMember.fromJson(Map<String, dynamic> json) {
    return TmdbCastMember(
      id: json['id'] ?? 0,
      name: json['name'] ?? '',
      character: json['character'],
      profilePath: json['profile_path'],
    );
  }

  String? get profileUrl => profilePath != null
      ? 'https://image.tmdb.org/t/p/w185$profilePath'
      : null;
}

/// Crew member from TMDB credits
class TmdbCrewMember {
  final int id;
  final String name;
  final String? job;
  final String? department;
  final String? profilePath;

  TmdbCrewMember({
    required this.id,
    required this.name,
    this.job,
    this.department,
    this.profilePath,
  });

  factory TmdbCrewMember.fromJson(Map<String, dynamic> json) {
    return TmdbCrewMember(
      id: json['id'] ?? 0,
      name: json['name'] ?? '',
      job: json['job'],
      department: json['department'],
      profilePath: json['profile_path'],
    );
  }

  String? get profileUrl => profilePath != null
      ? 'https://image.tmdb.org/t/p/w185$profilePath'
      : null;
}

/// Enriched metadata combining TMDB data
class TmdbEnrichedMetadata {
  final int id;
  final String? imdbId;
  final String title;
  final String? overview;
  final List<String> genres;
  final List<TmdbCastMember> cast;
  final List<TmdbCrewMember> crew;
  final String? rating;
  final String? posterPath;
  final String? backdropPath;
  final String? releaseDate;
  final int? runtime;
  final double? voteAverage;
  final List<TrailerInfo> trailers;

  TmdbEnrichedMetadata({
    required this.id,
    this.imdbId,
    required this.title,
    this.overview,
    this.genres = const [],
    this.cast = const [],
    this.crew = const [],
    this.rating,
    this.posterPath,
    this.backdropPath,
    this.releaseDate,
    this.runtime,
    this.voteAverage,
    this.trailers = const [],
  });

  String? get posterUrl => posterPath != null
      ? 'https://image.tmdb.org/t/p/w500$posterPath'
      : null;

  String? get backdropUrl => backdropPath != null
      ? 'https://image.tmdb.org/t/p/original$backdropPath'
      : null;

  TrailerInfo? get mainTrailer => trailers.isNotEmpty ? trailers.first : null;

  String get genresString => genres.join(', ');

  String get castString => cast.map((c) => c.name).join(', ');

  String? get director {
    final directors = crew.where((c) => c.job == 'Director');
    return directors.isNotEmpty ? directors.first.name : null;
  }

  int? get releaseYear {
    if (releaseDate == null) return null;
    final parts = releaseDate!.split('-');
    if (parts.isNotEmpty) {
      return int.tryParse(parts.first);
    }
    return null;
  }
}

File: OpenFlix-iOS/OpenFlix-iOS/Core/Network/TMDBService.swift
import Foundation

// MARK: - TMDB Service
// Fetches movie trailers from The Movie Database API
// API key is fetched from the OpenFlix server

actor TMDBService {
    static let shared = TMDBService()

    // TMDB API key - fetched from server
    private var apiKey: String?
    private var apiKeyLoaded = false
    private var apiKeyLoadTask: Task<Void, Never>?
    private let baseURL = "https://api.themoviedb.org/3"

    private let session: URLSession
    private let decoder: JSONDecoder

    // Cache trailers to avoid repeated API calls
    private var trailerCache: [String: TrailerInfo?] = [:]

    private init() {
        let config = URLSessionConfiguration.default
        config.timeoutIntervalForRequest = 10
        self.session = URLSession(configuration: config)
        self.decoder = JSONDecoder()
        self.decoder.keyDecodingStrategy = .convertFromSnakeCase
    }

    // MARK: - Public API

    /// Fetch trailer for a movie by TMDB ID
    func getTrailer(tmdbId: String) async -> TrailerInfo? {
        // Check cache first
        if let cached = trailerCache[tmdbId] {
            return cached
        }

        // Ensure API key is loaded from server
        await ensureApiKeyLoaded()

        // Skip if no API key available
        guard let apiKey = apiKey, !apiKey.isEmpty else {
            NSLog("TMDBService: No API key available from server")
            return nil
        }

        do {
            let trailer = try await fetchTrailer(tmdbId: tmdbId, apiKey: apiKey)
            trailerCache[tmdbId] = trailer
            return trailer
        } catch {
            NSLog("TMDBService: Failed to fetch trailer for \(tmdbId): \(error)")
            trailerCache[tmdbId] = nil
            return nil
        }
    }

    /// Load API key from server (called automatically when needed)
    private func ensureApiKeyLoaded() async {
        // If already loaded, return immediately
        if apiKeyLoaded {
            return
        }

        // If there's already a loading task, wait for it
        if let existingTask = apiKeyLoadTask {
            await existingTask.value
            return
        }

        // Create a new loading task
        let task = Task<Void, Never> {
            do {
                let response = try await OpenFlixAPI.shared.getServerSettings()
                self.apiKey = response.settings.tmdbApiKey
                if let key = self.apiKey, !key.isEmpty {
                    NSLog("TMDBService: Loaded TMDB API key from server: %@....", String(key.prefix(8)))
                } else {
                    NSLog("TMDBService: Server has no TMDB API key configured")
                }
            } catch {
                NSLog("TMDBService: Failed to load API key from server: \(error.localizedDescription)")
            }
            self.apiKeyLoaded = true
        }
        apiKeyLoadTask = task
        await task.value
    }

    /// Force reload API key from server (e.g., after settings change)
    func reloadApiKey() async {
        apiKeyLoaded = false
        apiKeyLoadTask = nil
        await ensureApiKeyLoaded()
    }

    /// Extract TMDB ID from Plex GUID
    /// Plex stores GUIDs like "plex://movie/5d776..." or "tmdb://12345" or "com.plexapp.agents.themoviedb://12345"
    func extractTMDBId(from guid: String?) -> String? {
        guard let guid = guid else { return nil }

        // Direct TMDB ID: "tmdb://12345"
        if guid.hasPrefix("tmdb://") {
            return String(guid.dropFirst("tmdb://".count))
        }

        // Legacy agent format: "com.plexapp.agents.themoviedb://12345?lang=en"
        if guid.contains("themoviedb://") {
            if let range = guid.range(of: "themoviedb://") {
                let afterPrefix = guid[range.upperBound...]
                // Extract ID before any query params
                let id = afterPrefix.split(separator: "?").first.map(String.init) ?? String(afterPrefix)
                return id
            }
        }

        return nil
    }

    /// Search TMDB for a movie by title and year, returns the TMDB ID if found
    func searchMovie(title: String, year: Int?) async -> String? {
        await ensureApiKeyLoaded()

        guard let apiKey = apiKey, !apiKey.isEmpty else {
            NSLog("TMDBService: Cannot search - no API key available")
            return nil
        }

        NSLog("TMDBService: Searching for movie '\(title)' (year: \(year ?? 0))")

        // Build search URL
        var urlString = "\(baseURL)/search/movie?api_key=\(apiKey)&query=\(title.addingPercentEncoding(withAllowedCharacters: .urlQueryAllowed) ?? title)"
        if let year = year {
            urlString += "&year=\(year)"
        }

        guard let url = URL(string: urlString) else {
            return nil
        }

        do {
            let (data, response) = try await session.data(from: url)

            guard let httpResponse = response as? HTTPURLResponse,
                  httpResponse.statusCode == 200 else {
                return nil
            }

            let searchResponse = try decoder.decode(TMDBSearchResponse.self, from: data)

            // Return the first result's ID
            if let firstResult = searchResponse.results.first {
                NSLog("TMDBService: Found TMDB match for '\(title)': ID \(firstResult.id)")
                return String(firstResult.id)
            }
        } catch {
            NSLog("TMDBService: Search failed for '\(title)': \(error)")
        }

        return nil
    }

    /// Get trailer by searching for movie title (fallback when no GUID)
    func getTrailerByTitle(title: String, year: Int?) async -> TrailerInfo? {
        // Search for the movie first
        guard let tmdbId = await searchMovie(title: title, year: year) else {
            NSLog("TMDBService: No TMDB match found for '\(title)'")
            return nil
        }

        // Then get the trailer
        return await getTrailer(tmdbId: tmdbId)
    }

    // MARK: - Private API

    private func fetchTrailer(tmdbId: String, apiKey: String) async throws -> TrailerInfo? {
        // Fetch movie details to get backdrop path
        let backdropPath = await fetchBackdropPath(tmdbId: tmdbId, apiKey: apiKey)

        let urlString = "\(baseURL)/movie/\(tmdbId)/videos?api_key=\(apiKey)"

        guard let url = URL(string: urlString) else {
            throw TMDBError.invalidURL
        }

        let (data, response) = try await session.data(from: url)

        guard let httpResponse = response as? HTTPURLResponse,
              httpResponse.statusCode == 200 else {
            throw TMDBError.requestFailed
        }

        let videosResponse = try decoder.decode(TMDBVideosResponse.self, from: data)

        // Find the best trailer: prefer official trailers from YouTube
        let trailers = videosResponse.results.filter { video in
            video.site.lowercased() == "youtube" &&
            video.type.lowercased() == "trailer"
        }

        // Prefer official trailers
        if let official = trailers.first(where: { $0.official == true }) {
            return TrailerInfo(youtubeKey: official.key, name: official.name, backdropPath: backdropPath)
        }

        // Fall back to any trailer
        if let trailer = trailers.first {
            return TrailerInfo(youtubeKey: trailer.key, name: trailer.name, backdropPath: backdropPath)
        }

        // Try teasers if no trailers available
        if let teaser = videosResponse.results.first(where: {
            $0.site.lowercased() == "youtube" && $0.type.lowercased() == "teaser"
        }) {
            return TrailerInfo(youtubeKey: teaser.key, name: teaser.name, backdropPath: backdropPath)
        }

        return nil
    }

    /// Fetch movie details to get the backdrop path
    private func fetchBackdropPath(tmdbId: String, apiKey: String) async -> String? {
        let urlString = "\(baseURL)/movie/\(tmdbId)?api_key=\(apiKey)"

        guard let url = URL(string: urlString) else { return nil }

        do {
            let (data, response) = try await session.data(from: url)
            guard let httpResponse = response as? HTTPURLResponse,
                  httpResponse.statusCode == 200 else { return nil }

            let movieDetails = try decoder.decode(TMDBMovieDetails.self, from: data)
            return movieDetails.backdropPath
        } catch {
            NSLog("TMDBService: Failed to fetch backdrop for \(tmdbId): \(error)")
            return nil
        }
    }

    /// Clear the trailer cache
    func clearCache() {
        trailerCache.removeAll()
    }

    // MARK: - Full Movie Details

    /// Fetch complete movie details from TMDB (for movies without server metadata)
    func getMovieDetails(title: String, year: Int?) async -> TMDBMovieInfo? {
        await ensureApiKeyLoaded()

        guard let apiKey = apiKey, !apiKey.isEmpty else {
            NSLog("TMDBService: Cannot fetch movie details - no API key available")
            return nil
        }

        // First search for the movie
        guard let tmdbId = await searchMovie(title: title, year: year) else {
            NSLog("TMDBService: No TMDB match found for '\(title)'")
            return nil
        }

        // Fetch full details with credits
        return await fetchMovieDetails(tmdbId: tmdbId, apiKey: apiKey)
    }

    /// Fetch movie details including cast, genres, and full overview
    private func fetchMovieDetails(tmdbId: String, apiKey: String) async -> TMDBMovieInfo? {
        // Use append_to_response to get credits in one call
        let urlString = "\(baseURL)/movie/\(tmdbId)?api_key=\(apiKey)&append_to_response=credits"

        guard let url = URL(string: urlString) else { return nil }

        do {
            let (data, response) = try await session.data(from: url)
            guard let httpResponse = response as? HTTPURLResponse,
                  httpResponse.statusCode == 200 else { return nil }

            let details = try decoder.decode(TMDBFullMovieDetails.self, from: data)

            // Convert to our simpler model
            return TMDBMovieInfo(
                tmdbId: details.id,
                title: details.title,
                overview: details.overview,
                backdropPath: details.backdropPath,
                posterPath: details.posterPath,
                releaseDate: details.releaseDate,
                runtime: details.runtime,
                voteAverage: details.voteAverage,
                genres: details.genres?.map { $0.name } ?? [],
                cast: details.credits?.cast?.prefix(10).map { castMember in
                    TMDBCastMember(
                        name: castMember.name,
                        character: castMember.character,
                        profilePath: castMember.profilePath
                    )
                } ?? [],
                directors: details.credits?.crew?.filter { $0.job == "Director" }.map { $0.name } ?? [],
                writers: details.credits?.crew?.filter { $0.job == "Writer" || $0.job == "Screenplay" }.map { $0.name } ?? []
            )
        } catch {
            NSLog("TMDBService: Failed to fetch movie details for \(tmdbId): \(error)")
            return nil
        }
    }
}

// MARK: - TMDB Response Models

struct TMDBSearchResponse: Codable {
    let page: Int
    let results: [TMDBSearchResult]
    let totalResults: Int?
    let totalPages: Int?

    enum CodingKeys: String, CodingKey {
        case page, results
        case totalResults = "total_results"
        case totalPages = "total_pages"
    }
}

struct TMDBSearchResult: Codable {
    let id: Int
    let title: String
    let originalTitle: String?
    let releaseDate: String?
    let posterPath: String?
    let backdropPath: String?
    let overview: String?

    enum CodingKeys: String, CodingKey {
        case id, title, overview
        case originalTitle = "original_title"
        case releaseDate = "release_date"
        case posterPath = "poster_path"
        case backdropPath = "backdrop_path"
    }
}

struct TMDBMovieDetails: Codable {
    let id: Int
    let title: String
    let backdropPath: String?
    let posterPath: String?
    let overview: String?

    enum CodingKeys: String, CodingKey {
        case id, title, overview
        case backdropPath = "backdrop_path"
        case posterPath = "poster_path"
    }
}

// Full movie details with credits
struct TMDBFullMovieDetails: Codable {
    let id: Int
    let title: String
    let overview: String?
    let backdropPath: String?
    let posterPath: String?
    let releaseDate: String?
    let runtime: Int?
    let voteAverage: Double?
    let genres: [TMDBGenre]?
    let credits: TMDBCredits?

    enum CodingKeys: String, CodingKey {
        case id, title, overview, runtime, genres, credits
        case backdropPath = "backdrop_path"
        case posterPath = "poster_path"
        case releaseDate = "release_date"
        case voteAverage = "vote_average"
    }
}

struct TMDBGenre: Codable {
    let id: Int
    let name: String
}

struct TMDBCredits: Codable {
    let cast: [TMDBCast]?
    let crew: [TMDBCrew]?
}

struct TMDBCast: Codable {
    let id: Int
    let name: String
    let character: String?
    let profilePath: String?
    let order: Int?

    enum CodingKeys: String, CodingKey {
        case id, name, character, order
        case profilePath = "profile_path"
    }
}

struct TMDBCrew: Codable {
    let id: Int
    let name: String
    let job: String
    let department: String?
    let profilePath: String?

    enum CodingKeys: String, CodingKey {
        case id, name, job, department
        case profilePath = "profile_path"
    }
}

// Simplified movie info for use in the app
struct TMDBMovieInfo {
    let tmdbId: Int
    let title: String
    let overview: String?
    let backdropPath: String?
    let posterPath: String?
    let releaseDate: String?
    let runtime: Int?
    let voteAverage: Double?
    let genres: [String]
    let cast: [TMDBCastMember]
    let directors: [String]
    let writers: [String]

    var backdropURL: URL? {
        guard let path = backdropPath else { return nil }
        return URL(string: "https://image.tmdb.org/t/p/w1280\(path)")
    }

    var posterURL: URL? {
        guard let path = posterPath else { return nil }
        return URL(string: "https://image.tmdb.org/t/p/w500\(path)")
    }
}

struct TMDBCastMember {
    let name: String
    let character: String?
    let profilePath: String?

    var profileURL: URL? {
        guard let path = profilePath else { return nil }
        return URL(string: "https://image.tmdb.org/t/p/w185\(path)")
    }
}

struct TMDBVideosResponse: Codable {
    let id: Int
    let results: [TMDBVideo]
}

struct TMDBVideo: Codable {
    let id: String
    let key: String  // YouTube video ID
    let name: String
    let site: String  // "YouTube", "Vimeo", etc.
    let type: String  // "Trailer", "Teaser", "Featurette", etc.
    let official: Bool?
    let publishedAt: String?

    enum CodingKeys: String, CodingKey {
        case id, key, name, site, type, official
        case publishedAt = "published_at"
    }
}

// MARK: - TMDB Errors

enum TMDBError: Error {
    case invalidURL
    case requestFailed
    case decodingError
    case noTrailerFound
}
